modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxUnlockRequestSender.java

```diff
@@ -0,0 +1,171 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.tx.impl;
+
+import static java.util.concurrent.CompletableFuture.allOf;
+import static org.apache.ignite.internal.hlc.HybridTimestamp.hybridTimestampToLong;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import org.apache.ignite.internal.hlc.HybridClock;
+import org.apache.ignite.internal.hlc.HybridTimestamp;
+import org.apache.ignite.internal.replicator.ReplicationGroupId;
+import org.apache.ignite.internal.replicator.TablePartitionId;
+import org.apache.ignite.internal.tx.impl.TxManagerImpl.TransactionFailureHandler;
+import org.apache.ignite.internal.tx.message.LockReleaseMessage;
+import org.apache.ignite.internal.tx.message.TxMessagesFactory;
+import org.apache.ignite.internal.util.CompletableFutures;
+import org.apache.ignite.network.ClusterService;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Sends TX Unlock request.
+ */
+public class TxUnlockRequestSender {
+
+    /** Network timeout. */
+    private static final long RPC_TIMEOUT = 3000;
+
+    /** Tx messages factory. */
+    private static final TxMessagesFactory FACTORY = new TxMessagesFactory();
+
+    /** Cluster service. */
+    private final ClusterService clusterService;
+
+    /** Placement driver helper. */
+    private final PlacementDriverHelper placementDriverHelper;
+
+    /** Hybrid clock. */
+    private final HybridClock hybridClock;
+
+    /** Cleanup processor. */
+    private final TxCleanupProcessor txCleanupProcessor;
+
+    /**
+     * The constructor.
+     *
+     * @param clusterService Cluster service.
+     * @param placementDriverHelper Placement driver helper.
+     * @param clock A hybrid logical clock.
+     * @param txCleanupProcessor A cleanup processor.
+     */
+    public TxUnlockRequestSender(
+            ClusterService clusterService,
+            PlacementDriverHelper placementDriverHelper,
+            HybridClock clock,
+            TxCleanupProcessor txCleanupProcessor) {
+        this.clusterService = clusterService;
+        this.placementDriverHelper = placementDriverHelper;
+        this.hybridClock = clock;
+        this.txCleanupProcessor = txCleanupProcessor;
+    }
+
+    /**
+     * Sends unlock request to the primary nodes of each one of {@code partitions}.
+     *
+     * @param partitions Enlisted partition groups.
+     * @param commit {@code true} if a commit requested.
+     * @param commitTimestamp Commit timestamp ({@code null} if it's an abort).
+     * @param txId Transaction id.
+     * @return Completable future of Void.
+     */
+    public CompletableFuture<Void> unlock(
```

Comment:
Naming is still a bit messy. And thus same for components areas of responsibilities - it's also messy. On the one hand we consider cleanup as an umbrella operations that consists of unlock and write intents conversion, on the other hand unlock within given method calls durable cleanup, so it's reversed. 

From my point of view, first approach is better. In order not to mix writeIntent cleanup with general cleanup that also includes unlock I'd rather introduce new term - writeIntentSwitch.